core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1387,16 +1387,18 @@ private[spark] object Utils extends Logging {
           catchBlock
         } catch {
           case t: Throwable =>
-            originalThrowable.addSuppressed(t)
-            logWarning(s"Suppressing exception in catch: " + t.getMessage, t)
+            if (originalThrowable != t) {
+              originalThrowable.addSuppressed(t)
+              logWarning(s"Suppressing exception in catch: " + t.getMessage, t)
+            }
         }
         throw originalThrowable
     } finally {
       try {
         finallyBlock
       } catch {
         case t: Throwable =>
-          if (originalThrowable != null) {
+          if (originalThrowable != null && originalThrowable != t) {
```

Comment:
Really in this case and the similar one at the top you could write:

```
case t: Throwable if (originalThrowable != null && originalThrowable != t) =>
```

because otherwise it's just re-throwing the exception anyway rather than handling it